### PR TITLE
Module Badges Fixture

### PR DESCRIPTION
**Pull request by esoergel**
## Technical Summary
https://docs.google.com/document/d/1rdSZfaMH-qjaTNlUHdOjWCmkcPL4bRdU7iSP5OYfs7c/edit
Base branch from https://github.com/dimagi/commcare-hq/pull/35299
This takes the configuration provided by that PR and turns it into a fixture in the restore, with a small custom component for BHA

## Feature Flag
MODULE_BADGES

## Safety Assurance
Fully behind a feature flag, also tested on staging

### Rollback instructions
- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

### CODE REVIEW

File: corehq/apps/case_search/fixtures.py [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+    return {
+        "username": restore_user.username,
+        "uuid": restore_user.user_id,
+        "user_data": restore_user.user_session_data
+    }
+
+
+def _get_template_renderer(restore_user):
+    renderer = MessagingTemplateRenderer()
+    renderer.set_context_param('user', NestedDictTemplateParam(_get_user_template_info(restore_user)))
```
Comment (esoergel) [inline corehq/apps/case_search/fixtures.py:24]: This is where the context is provided for interpolating the CSQL template strings.  App builders will only be able to reference things we add here, which as of this writing is only username, ID, and user data

File: corehq/apps/case_search/filter_dsl.py [hunk 2]
```diff
@@ -149,6 +149,10 @@ def build_filter_from_xpath(xpath, *, domain=None, context=None):
             raise CaseFilterError(error_message.format(bad_part, ", ".join(ALL_OPERATORS)), bad_part)
         raise CaseFilterError(_("Malformed search query"), None)
     except RuntimeError as e:
+        # eulxml doesn't appear to clean up after this type of failure
+        # properly, so throw in an extra 'parse' to reset it
+        parse_xpath("thisisdumb")
```
Comment (esoergel) [inline corehq/apps/case_search/filter_dsl.py:154]: Really the fix here is for us to get off of `eulxml`, which we've wanted to do a couple times in the past as well.  Just made a backlog item for that:
https://dimagi.atlassian.net/browse/USH-5070

File: corehq/apps/case_search/fixtures.py [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+    for name, param in custom_csql_fixture_context(restore_user.domain, restore_user):
+        renderer.set_context_param(name, param)
+    return renderer
+
+
+@extensions.extension_point
```
[the remark this answers is not in the document]
Comment (esoergel) [reply corehq/apps/case_search/fixtures.py:35]: Yeah, I found it to be a really nice fit for this use-case (in my opinion, anyways)

File: corehq/apps/case_search/fixtures.py [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+def custom_csql_fixture_context(domain, restore_user):
+    '''Register custom template params to be available in CSQL templates'''
```
Comment (MartinRiese) [inline corehq/apps/case_search/fixtures.py:37]: A reference to `corehq/apps/app_manager/suite_xml/post_process/instances.py` would be nice for future devs reading this code.

File: corehq/apps/case_search/fixtures.py [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+               .filter(filter_)
+               .count())
+
+
+def _get_indicator_nodes(restore_state, indicators):
+    renderer = _get_template_renderer(restore_state.restore_user)
+    for name, csql_template in indicators:
+        with restore_state.timing_context(name):
+            value = _run_query(restore_state.domain, renderer.render(csql_template))
+        yield E.value(value, name=name)
```
[the remark this answers is not in the document]
Comment (MartinRiese) [reply corehq/apps/case_search/fixtures.py:56]: We could add a test button to the view.

File: custom/bha/commcare_extensions.py [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+from .util import get_user_clinic_ids, get_user_facility_ids
+
+BHA_DOMAINS = settings.CUSTOM_DOMAINS_BY_MODULE['custom.bha']
+
+
+@custom_csql_fixture_context.extend(domains=BHA_DOMAINS)
+def bha_csql_fixture_context(domain, restore_user):
+    facility_ids = get_user_facility_ids(domain, restore_user)
+    return ('bha', SimpleDictTemplateParam({
+        'facility_ids': facility_ids,
```
Comment (MartinRiese) [inline custom/bha/commcare_extensions.py:15]: Does this mean the facility_ids are evaluated every time even when they are not used in the CSQL?

File: corehq/messaging/templating.py [hunk 7]
```diff
@@ -142,6 +142,22 @@ def __getattr__(self, item):
         return SimpleMessagingTemplateParam(UNKNOWN_VALUE)
 
 
+class NestedDictTemplateParam(SimpleDictTemplateParam):
+
+    def __init__(self, dict_of_values):
+        self.__dict_of_values = dict_of_values
```
[the remark this answers is not in the document]
Comment (esoergel) [reply corehq/messaging/templating.py:148]: Actually yes - this was another super cryptic issue I ran into.  I think it has something to do with overriding `__getattr__` and method resolution order.  Without this, you get an error about the class not having an attribute called `__dict_of_values`

File: corehq/apps/case_search/fixtures.py (hunk 4 above)
Comment (esoergel) [reply corehq/apps/case_search/fixtures.py:37]: Can you elaborate?  They shouldn't need to register anything else there if extending this work.  I did consider importing `CaseSearchFixtureProvider` over there to reference the `id`, but I see we don't do that for the other fixtures, and I didn't want to introduce weird import dependencies.

File: custom/bha/commcare_extensions.py (hunk 6 above)
Comment (esoergel) [reply custom/bha/commcare_extensions.py:15]: Yep, though only for BHA.  The fixture context is generated once per restore though, so if any indicators in the fixture use it, it's not wasted effort.  If we think this param (or future ones) aren't actually needed all the time, we could probably write a custom template param that lazily fetches expensive fields on demand.

File: corehq/apps/case_search/fixtures.py (hunk 4 above)
Comment (MartinRiese) [reply corehq/apps/case_search/fixtures.py:37]: Copied the link for the wrong file. I meant the settings. But thinking more about it, it is pretty easy to follow the names to get there.

**Review by MartinRiese — DISMISSED** [review]

**Review by AddisonDunn — DISMISSED** [review]

**Review by AddisonDunn — APPROVED** [review]

**Pull request merged**